File: backend/src/main/java/com/taskflow/domain/Board.java
package com.taskflow.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

/**
 * 보드(컬렉션) 엔티티
 *
 * 테이블: TB_BOARD
 *
 * 업무 아이템들을 담는 컬렉션(보드)
 * - 소유자가 보드를 생성하고 관리
 * - 다른 사용자에게 공유 가능
 */
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Board {

    /**
     * 보드 ID (PK)
     */
    private Long boardId;

    /**
     * 보드명
     */
    private String boardName;

    /**
     * 보드 설명
     */
    private String description;

    /**
     * 소유자 ID (FK)
     */
    private Long ownerId;

    /**
     * 기본 뷰 타입 (TABLE, KANBAN, LIST)
     */
    private String defaultView;

    /**
     * 표시 색상 (#RRGGBB)
     */
    private String color;

    /**
     * 정렬 순서
     */
    private Integer sortOrder;

    /**
     * 사용 여부 (Y/N)
     */
    private String useYn;

    /**
     * 생성일시
     */
    private LocalDateTime createdAt;

    /**
     * 생성자 ID
     */
    private Long createdBy;

    /**
     * 수정일시
     */
    private LocalDateTime updatedAt;

    /**
     * 수정자 ID
     */
    private Long updatedBy;

    // =============================================
    // 추가 필드 (Mapper에서 설정)
    // =============================================

    /**
     * 소유자명
     */
    private String ownerName;

    /**
     * 아이템 수
     */
    private Integer itemCount;

    /**
     * 공유 사용자 수
     */
    private Integer shareCount;

    // =============================================
    // 편의 메서드
    // =============================================

    /**
     * 활성 보드 여부
     */
    public boolean isActive() {
        return "Y".equals(useYn);
    }

    /**
     * 소유자 여부 확인
     */
    public boolean isOwner(Long userId) {
        return ownerId != null && ownerId.equals(userId);
    }
}

File: frontend/src/types/board.ts
/**
 * 보드(컬렉션) 타입 정의
 */

export type ViewType = 'TABLE' | 'KANBAN' | 'LIST'

export interface BoardGroup {
  groupId: number
  groupCode: string
  groupName: string
  description?: string
  color?: string
  sortOrder?: number
  useYn?: string
  memberCount?: number
}

export interface BoardSharedUser {
  username: string      // 로그인 ID (백엔드 FK 참조용)
  userName: string      // 표시용 이름
  departmentName?: string
}

export interface Board {
  boardId: number
  boardName: string
  boardDescription?: string
  description?: string  // API 응답 호환
  ownerUsername?: string   // 소유자 USERNAME
  ownerName?: string       // 소유자 표시명
  defaultView: ViewType
  boardColor?: string
  color?: string  // API 응답 호환
  shareCount: number
  itemCount?: number
  pendingItemCount?: number
  sortOrder?: number
  useYn?: string
  ownerYn?: string
  currentUserPermission?: BoardPermission
  isOwner?: boolean
  groups?: BoardGroup[]
  sharedUsers?: BoardSharedUser[]
  createdAt: string
  updatedAt?: string
}

export interface BoardCreateRequest {
  boardName: string
  boardDescription?: string
  defaultView?: ViewType
  color?: string
}

export interface BoardUpdateRequest {
  boardName?: string
  boardDescription?: string
  defaultView?: ViewType
  color?: string
}

export type BoardPermission = 'VIEW' | 'EDIT' | 'FULL' | 'OWNER'

export interface BoardShare {
  boardShareId: number
  boardId: number
  boardName?: string
  username: string        // FK 참조용 USERNAME
  userName?: string       // 표시용 이름
  departmentName?: string
  permission: BoardPermission
  createdAt: string
}

export interface BoardShareRequest {
  username: string  // 백엔드 API는 username 필요
  permission: BoardPermission
}

export interface BoardShareUpdateRequest {
  permission: BoardPermission
}

export interface BoardListResponse {
  ownedBoards: Board[]
  sharedBoards: Board[]
  totalOwnedCount: number
  totalSharedCount: number
}

export interface BoardDeleteRequest {
  targetUsername?: string   // 이관 대상 사용자 USERNAME
  forceDelete?: boolean
}

export interface BoardOrderRequest {
  sortOrder: number
}

export interface TransferPreviewResponse {
  boardId: number
  boardName: string
  pendingItems: PendingItem[]
  totalCount: number
}

export interface PendingItem {
  itemId: number
  title: string
  status: string
  priority?: string
  assigneeName?: string
}

export interface TransferResultResponse {
  transferredCount: number
  newBoardId: number
  newBoardName: string
  message: string
}

/**
 * 보드 소유권 이전 요청 DTO
 */
export interface BoardTransferRequest {
  targetUsername: string  // 이관받을 사용자 USERNAME
  reason?: string         // 이관 사유 (선택)
}
